### PR TITLE
Handle rented node in grace period

### PR DESCRIPTION
**Pull request by samaradel**
### Description

Check if the rented node contract is in the grace period when listing the node.

### Changes

image

### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3025

### Tested Scenarios

- Navigate to the Node finder page.
- Rent a dedicated node.
- Leave it till it goes to the grace period.
- Navigate to any solution.
- Try to deploy an instance on the rented node.

### Checklist

- [x] Tests included
- [x] Build pass
- [ ] Documentation
- [x] Code format and docstrings
- [x] Screenshots/Video attached (needed for UI changes)

### CODE REVIEW

**Review by 0oM4R — COMMENTED** [review]
image


i see two toasts appears, should be only on and should be inside the node finder

**Review by 0oM4R — COMMENTED** [review]
still having toast 

[Screencast from 2024-11-20 16-27-08.webm](https://github.com/user-attachments/assets/6b72b11b-4781-420b-b1bf-20993a39cd84)

also the inside node selector the error is shown as info in  first

**Review by ramezsaeed — APPROVED** [review]

File: packages/playground/src/components/node_selector/TfAutoNodeSelector.vue [hunk 1]
```diff
@@ -151,6 +151,9 @@ import { computed, nextTick, onMounted, onUnmounted, type PropType, ref } from "
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
 import type { VCard } from "vuetify/components/VCard";
 
+import { createCustomToast, ToastType } from "@/utils/custom_toast";
```
Comment (0oM4R) [inline packages/playground/src/components/node_selector/TfAutoNodeSelector.vue]: are those used anywhere?

File: packages/playground/src/components/node_selector/TfAutoNodeSelector.vue [hunk 2]
```diff
@@ -299,14 +301,23 @@ export default {
...
-          });
-          if (state.gracePeriod) {
-            return false;
+        try {
+          if (node && node.rentContractId !== 0) {
+            const { state } = await gridStore.grid.contracts.get({
+              id: node?.rentContractId,
+            });
+            if (state.gracePeriod) {
+              throw `You can't deploy on node ${node.nodeId}, its rent contract is in grace period.`;
```
Comment (0oM4R) [inline packages/playground/src/components/node_selector/TfAutoNodeSelector.vue]: i suggest adding time out before throwing the error, for more friendly UX.

File: packages/playground/src/utils/nodeSelector.ts [hunk 3]
```diff
@@ -229,10 +229,7 @@ export async function validateRentContract(
         id: node.rentContractId,
       });
       if (contractInfo.state.gracePeriod) {
-        createCustomToast(
-          `You can't deploy on node ${node.nodeId}, its rent contract is in grace period.`,
-          ToastType.danger,
-        );
+        throw `You can't deploy on node ${node.nodeId}, its rent contract is in grace period.`;
```
Comment (0oM4R) [inline packages/playground/src/utils/nodeSelector.ts]: i don't think we should throw an error here, we should update the field message. 
image

Comment (0oM4R) [reply packages/playground/src/utils/nodeSelector.ts]: image

File: packages/playground/src/components/node_selector/TfAutoNodeSelector.vue (hunk 1 above)
Comment (samaradel) [reply packages/playground/src/components/node_selector/TfAutoNodeSelector.vue]: checked the associated PRs and I see it's not anymore.

File: packages/playground/src/components/node_selector/TfAutoNodeSelector.vue (hunk 2 above)
Comment (samaradel) [reply packages/playground/src/components/node_selector/TfAutoNodeSelector.vue]: why do we need timeout here? we check the state object only.

File: packages/playground/src/utils/nodeSelector.ts (hunk 3 above)
Comment (samaradel) [reply packages/playground/src/utils/nodeSelector.ts]: I don't think so as it's doing the double toast issue and it all should take the same behavior (throwing an error or firing a toast)

File: packages/playground/src/components/node_selector/TfAutoNodeSelector.vue (hunk 2 above)
Comment (0oM4R) [reply packages/playground/src/components/node_selector/TfAutoNodeSelector.vue]: https://github.com/threefoldtech/tfgrid-sdk-ts/pull/3646#pullrequestreview-2448787416
in this video, the switching between info and error toast is to fast, if you have any other solutoin to hide the info toast then show the error will be much better imo

[Screencast from 2024-12-01 11-27-36.webm](https://github.com/user-attachments/assets/25278603-d7a4-4848-9299-9cf5b2d7d6d0)

File: packages/playground/src/utils/nodeSelector.ts (hunk 3 above)
Comment (0oM4R) [reply packages/playground/src/utils/nodeSelector.ts]: Sorry, I disagree. I don't see any validation error in the manual selector that appears on toast, all errors are shown in the field info, As you said, we need to make the behavior consistent. @AhmedHanafy725  what do you think?

Comment (AhmedHanafy725) [reply packages/playground/src/utils/nodeSelector.ts]: It should be consistent. The error shouldn't be in the toast, it should be a field error

File: packages/playground/src/components/node_selector/TfAutoNodeSelector.vue (hunk 2 above)
Comment (samaradel) [reply packages/playground/src/components/node_selector/TfAutoNodeSelector.vue]: fixed

**Comment by amiraabouhadid** [comment]
no toast error, only console err
image
not sure if resources issue is from this pr changes or in dashboard overall
field error takes a while until it shows
image

File: packages/playground/src/components/node_selector/TfAutoNodeSelector.vue [hunk 4]
```diff
@@ -299,14 +298,27 @@ export default {
...
-            return false;
+        try {
+          if (node && node.rentContractId !== 0) {
+            const { state } = await gridStore.grid.contracts.get({
+              id: node?.rentContractId,
+            });
+            if (state.gracePeriod) {
+              await new Promise((_, reject) => {
+                setTimeout(() => {
+                  reject(Error(`You can't deploy on node ${node.nodeId}, its rent contract is in grace period.`));
```
Comment (amiraabouhadid) [inline packages/playground/src/components/node_selector/TfAutoNodeSelector.vue:309]: shouldn't we use console.err?

File: packages/playground/src/components/node_selector/TfAutoNodeSelector.vue [hunk 5]
```diff
@@ -299,14 +298,27 @@ export default {
...
+                }, 2000);
+              });
+            }
           }
+        } catch (error) {
+          const err = normalizeError(
```
Comment (amiraabouhadid) [inline packages/playground/src/components/node_selector/TfAutoNodeSelector.vue:315]: shouldnt we use console.err instead ?

Comment (samaradel) [reply packages/playground/src/components/node_selector/TfAutoNodeSelector.vue:315]: Instead? I don't think so. It's a connection error, which the user should be aware of in toast.

**Review by 0oM4R — COMMENTED** [review]
on selecting the node with the automatic node selector, i got this error in console

image

**Pull request closed without merging**